Mathlib/Topology/ContinuousFunction/CompactlySupported.lean

```diff
@@ -0,0 +1,541 @@
+`C_c(α, β)` will not have a multiplicative identity.
+-/
+
+section AlgebraicStructure
+
+variable [TopologicalSpace β] (x : α)
+
+instance [Zero β] : Zero C_c(α, β) where
+  zero := { toFun := (0 : C(α, β))
+            continuous_toFun := (0 : C(α, β)).2
+            hasCompactSupport' := by simp [HasCompactSupport, tsupport] }
+
+instance [Zero β] : Inhabited C_c(α, β) :=
+  ⟨0⟩
+
+@[simp]
+theorem coe_zero [Zero β] : ⇑(0 : C_c(α, β)) = 0 :=
+  rfl
+
+theorem zero_apply [Zero β] : (0 : C_c(α, β)) x = 0 :=
+  rfl
+
+instance [MulZeroClass β] [ContinuousMul β] : Mul C_c(α, β) :=
+  ⟨fun f g => ⟨f * g, HasCompactSupport.mul_left g.2⟩⟩
+
+@[simp]
+theorem coe_mul [MulZeroClass β] [ContinuousMul β] (f g : C_c(α, β)) : ⇑(f * g) = f * g :=
+  rfl
+
+theorem mul_apply [MulZeroClass β] [ContinuousMul β] (f g : C_c(α, β)) : (f * g) x = f x * g x :=
+  rfl
+
+/-- the product of `f : F` assuming `ContinuousMapClass F α γ` and `ContinuousSMul γ β` and
+`g : C_c(α, β)` is in `C_c(α, β)` -/
+instance [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
+    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] : SMul F C_c(α, β) where
+  smul f g :=
+    ⟨⟨fun x ↦ f x • g x, (map_continuous f).smul g.continuous⟩, g.hasCompactSupport'.smul_left⟩
+
+@[simp]
+theorem coe_smulc [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
+    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] (f : F) (g : C_c(α, β)) :
+    ⇑(f • g) = fun x => f x • g x :=
+  rfl
+
+@[simp]
+theorem smulc_apply [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
+    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] (f : F) (g : C_c(α, β)) (x : α):
+    (f • g) x = f x • g x :=
+  rfl
+
+instance [MulZeroClass β] [ContinuousMul β] : MulZeroClass C_c(α, β) :=
+  DFunLike.coe_injective.mulZeroClass _ coe_zero coe_mul
+
+instance [SemigroupWithZero β] [ContinuousMul β] :
+    SemigroupWithZero C_c(α, β) :=
+  DFunLike.coe_injective.semigroupWithZero _ coe_zero coe_mul
+
+instance [AddZeroClass β] [ContinuousAdd β] : Add C_c(α, β) :=
+  ⟨fun f g => ⟨f + g, HasCompactSupport.add f.2 g.2⟩⟩
+
+@[simp]
+theorem coe_add [AddZeroClass β] [ContinuousAdd β] (f g : C_c(α, β)) : ⇑(f + g) = f + g :=
+  rfl
+
+theorem add_apply [AddZeroClass β] [ContinuousAdd β] (f g : C_c(α, β)) : (f + g) x = f x + g x :=
+  rfl
+
+instance [AddZeroClass β] [ContinuousAdd β] : AddZeroClass C_c(α, β) :=
+  DFunLike.coe_injective.addZeroClass _ coe_zero coe_add
+
+/-- Coercion to a function as a `AddMonoidHom`. Similar to `AddMonoidHom.coeFn`. -/
+def coeFnMonoidHom [AddMonoid β] [ContinuousAdd β] : C_c(α, β) →+ α → β where
+  toFun f := f
+  map_zero' := coe_zero
+  map_add' := coe_add
+
+instance [Zero β] {R : Type*} [Zero R] [SMulWithZero R β] [ContinuousConstSMul R β] :
+    SMul R C_c(α, β) :=
+  ⟨fun r f => ⟨⟨r • ⇑f, Continuous.const_smul f.continuous r⟩, HasCompactSupport.smul_left' f.2⟩⟩
```

Comment:
```suggestion
instance [Zero β] {R : Type*} [SMulZeroClass R β] [ContinuousConstSMul R β] :
    SMul R C_c(α, β) :=
  ⟨fun r f => ⟨⟨r • ⇑f, Continuous.const_smul f.continuous r⟩, HasCompactSupport.smul_left f.2⟩⟩
```
No need for `SMulWithZero`, you can do with the weaker `[SMulZeroClass R β]`. Can you fix the next lemmas acordingly?

Reply:
done.